### PR TITLE
Adds a new normalizer to normalize œ to oe and æ to ae

### LINKED ISSUE #268: Normalize "œ" / "æ" into "oe" / "ae"
Implement a normalizer normalizing "œ" / "æ" into "oe" / "ae".

See the [Normalizer template](https://github.com/meilisearch/charabia/blob/main/charabia/src/normalizer/dummy_example.rs) to fix this issue.

## Related

- https://github.com/meilisearch/meilisearch/issues/4405

--------------------------------

> Hey! 👋 
Before starting any implementation, make sure that you read the [CONTRIBUTING.md](https://github.com/meilisearch/charabia/blob/main/CONTRIBUTING.md#contributing) file.
In addition to the recurrent rules, you can find some guides to easily implement a `Segmenter` or a `Normalizer`.
Thanks a lot for your Contribution! 🤝

### PR DESCRIPTION
**Pull request by Soham1803**
# Pull Request

## Related issue
Fixes #268

## What does this PR do?
- Creates a new normalizer *ae_oe_normalizer*
- normalizes `œ` and `Œ`  to `oe`,  `æ` and `Æ` to `ae`. 

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
+// Import `CharNormalizer` trait.
+use super::{ CharNormalizer, CharOrStr };
+use crate::Token;
+
+// Make a small documentation of the specialized Normalizer like below.
+/// <Script/Language> specialized [`Normalizer`].
+///
+/// This Normalizer uses [`<UsedLibraryToNormalize>`] internally to normalize the provided token.
+/// <OptionalAdditionnalExplanations>
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:9]: Can you please modify the documentation in order to explain the implementation, please?

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                token.lemma.chars().any('Œ') ||
+                token.lemma.chars().any('Æ'))
+    }
+}
+
+// Include the newly implemented Normalizer in the tokenization pipeline:
+//     - change the name of the file `dummy_example.rs` to `dummy.rs`
+//     - import module by adding `mod dummy;` (filename) in `normalizer/mod.rs`
+//     - Add Normalizer in `NORMALIZERS` in `normalizer/mod.rs`
+//     - check if it didn't break any test or benhchmark
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:42]: The normalizer is implemented, however, you didn't import it in the mod.rs file and added in `NORMALIZERS`, you should follow the bellow instructions and remove them from the file:
```suggestion
```

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            'æ' => Some("ae".to_string().into()),
+            'Œ' => Some("OE".to_string().into()),
+            'Æ' => Some("AE".to_string().into()),
+            _ => Some(c.into()),
+        }
+    }
+
+    // Returns `true` if the Normalizer should be used.
+    fn should_normalize(&self, token: &Token) -> bool {
+        // here we lowercase only on Latin and Cyrillic Scripts and if the current token contains an uppercased character.
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:28]:
```suggestion
```

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                kind: TokenKind::Word,
+                ..Default::default()
+            }
+        ]
+    }
+
+    test_normalizer!(AE_OE_NormalizerNormalizer, tokens(), normalizer_result(), normalized_tokens());
+}
+
+// Your Normalizer will now be used on texts of the assigned Script and Language. Thank you for your contribution, and congratulation! 🎉
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:162]:
```suggestion
```

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+/// <Script/Language> specialized [`Normalizer`].
+///
+/// This Normalizer uses [`<UsedLibraryToNormalize>`] internally to normalize the provided token.
+/// <OptionalAdditionnalExplanations>
+pub struct OE_AE_Normalizer;
+
+// All normalizers only need to implement the method `normalize_char` and the method `should_normalize` of the `CharNormalizer` trait.
+impl CharNormalizer for AE_OE_Normalizer {
+    // Creates the normalized version of the provided char.
+    // In this example we will remove whitespaces and lowercase other characters.
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:15]:
```suggestion
```

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+use crate::Token;
+
+// Make a small documentation of the specialized Normalizer like below.
+/// <Script/Language> specialized [`Normalizer`].
+///
+/// This Normalizer uses [`<UsedLibraryToNormalize>`] internally to normalize the provided token.
+/// <OptionalAdditionnalExplanations>
+pub struct OE_AE_Normalizer;
+
+// All normalizers only need to implement the method `normalize_char` and the method `should_normalize` of the `CharNormalizer` trait.
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:12]:
```suggestion
```

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
+// Import `CharNormalizer` trait.
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:1]:
```suggestion
```

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+    // Returns `true` if the Normalizer should be used.
+    fn should_normalize(&self, token: &Token) -> bool {
+        // here we lowercase only on Latin and Cyrillic Scripts and if the current token contains an uppercased character.
+        token.script == Script::Latin &&
+            token.script == Script::Cyrillic &&
+            (token.lemma.chars().any('œ') ||
+                token.lemma.chars().any('æ') ||
+                token.lemma.chars().any('Œ') ||
+                token.lemma.chars().any('Æ'))
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs.rs:34]: This implementation will iterate several times on all characters of the string, it's not optimal, we should instead iterate on time:
```suggestion
            token.lemma.chars().any(|c| matches!(c, 'œ' | 'æ' | 'Œ' | 'Æ'))
```

**Review by ManyTheFish — CHANGES_REQUESTED** [review]

**Comment by ManyTheFish** [comment]
Hello @Soham1803,
Thank you for your PR, I made several change requests before accepting to merge it,

let me know if you need more informations

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs (hunk 2 above)
Comment (Soham1803) [reply charabia/src/normalizer/ae_oe_normalizer.rs.rs:42]: Sure, just want to confirm one thing, should it be added in `NORMALIZERS` or `LOSSY_NORMALIZERS`?

**Comment by Soham1803** [comment]
Hey @ManyTheFish I have made the requested changes can you please check and let me know. Thank You!

File: charabia/src/normalizer/ae_oe_normalizer.rs.rs (hunk 2 above)
Comment (ManyTheFish) [reply charabia/src/normalizer/ae_oe_normalizer.rs.rs]: Hello @Soham1803,

deeply sorry for the delay,
`LOSSY_NORMALIZER` is better

**Review by ManyTheFish — DISMISSED** [review]
Hello @Soham1803 

Thank you for the PR,
nice work,

bors merge

**Comment by ManyTheFish** [comment]
Hello @Soham1803,

The tests don't seem to work. Could you fix them?

Thank you

**Comment by Soham1803** [comment]
Hey @ManyTheFish, I'm trying to run the tests on my local machine but got some issues with `jemalloc-sys` in *.cargo*. The last commit I made is based on the logs of test run I read on *Github*. Its good if the tests pass now, or else I will make another commit after being assured all the tests are passed.

File: charabia/src/normalizer/ae_oe_normalizer.rs [hunk 9]
```diff
@@ -0,0 +1,154 @@
+use crate::normalizer::NormalizerOption;
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs:1]:
```suggestion
```

File: charabia/Cargo.toml [hunk 10]
```diff
@@ -33,6 +33,7 @@ unicode-normalization = "0.1.22"
 irg-kvariants = "0.1.0"
 litemap = "0.7.2"
 zerovec = "0.10.1"
+jemalloc-sys = "0.5.4"
```
Comment (ManyTheFish) [inline charabia/Cargo.toml:36]: This may be not usefull
```suggestion
```

File: charabia/src/normalizer/ae_oe_normalizer.rs [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+use super::{ CharNormalizer, CharOrStr };
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs:2]:
```suggestion
use super::{CharNormalizer, CharOrStr};
```

File: charabia/src/normalizer/ae_oe_normalizer.rs [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            _ => Some(c.into()),
+        }
+    }
+
+    // Returns `true` if the Normalizer should be used.
+    fn should_normalize(&self, token: &Token) -> bool {
+        
+        token.script == Script::Latin &&
+        token.script == Script::Cyrillic &&
+        token.lemma.chars().any(is_should_normalize)
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs:29]:
```suggestion
        token.script == Script::Latin
            && token.script == Script::Cyrillic
            && token.lemma.chars().any(is_should_normalize)
```

File: charabia/src/normalizer/ae_oe_normalizer.rs [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+                script: Script::Latin,
+                ..Default::default()
+            },
+            Token {
+                lemma: Owned("Æ".to_string()),
+                char_end: 1,
+                byte_end: 2,
+                script: Script::Latin,
+                ..Default::default()
+            }
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs:76]:
```suggestion
            },
```

File: charabia/src/normalizer/ae_oe_normalizer.rs [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+                script: Script::Latin,
+                ..Default::default()
+            },
+            Token {
+                lemma: Owned("AE".to_string()),
+                char_end: 2,
+                byte_end: 2,
+                script: Script::Latin,
+                ..Default::default()
+            }
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs:110]:
```suggestion
            },
```

File: charabia/src/normalizer/ae_oe_normalizer.rs [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+                ..Default::default()
+            },
+            Token {
+                lemma: Owned("ae".to_string()),
+                char_end: 2,
+                byte_end: 2,
+                script: Script::Latin,
+                kind: TokenKind::Word,
+                ..Default::default()
+            }
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs:148]:
```suggestion
            },
```

File: charabia/src/normalizer/ae_oe_normalizer.rs [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        ]
+    }
+
+    test_normalizer!(AeOeNormalizer, tokens(), normalizer_result(), normalized_tokens());
+}
+
```
Comment (ManyTheFish) [inline charabia/src/normalizer/ae_oe_normalizer.rs:154]:
```suggestion
```

**Review by ManyTheFish — CHANGES_REQUESTED** [review]
The changes that I requested should fix the errors to run the tests

**Comment by Soham1803** [comment]
Hey @ManyTheFish sorry for the delay. Thanks for your suggested changes. The last commit passed all the tests on local successfully, after many different tryouts. Made some major changes to the Normalizer. Suggest me any changes if necessary before the merge.

Thank You!

**Comment by curquiza** [comment]
Hello @Soham1803 

@ManyTheFish is on Holidays and will review your PR when coming back

In the meantime, can you fix the Rustfmt tests? 😊 

Thanks again for your PR

**Comment by Soham1803** [comment]
Hello @curquiza, dealt with the **Rustfmt** tests. I guess all tests are passed. Now, just waiting for @ManyTheFish to suggest any required changes before the merge.

Thank You! 😊

**Review by ManyTheFish — APPROVED** [review]
LGTM!
Thank you @Soham1803 for your work. This should be part of the Meilisearch v1.9.0 release!

bors merge

**Comment by Soham1803** [comment]
Nice to have you back @ManyTheFish. Thanks for accepting my PR and the help you provided. 😊

**Pull request merged**